### ISSUE TITLE
.equals returns None instead of bool

### ISSUE THREAD

**Issue by Karrenbelt** [sympy/sympy#18719]
expr1 = sympy.sympify('x*x', evaluate=False)
expr2 = sympy.sympify('x**2')
expr3 = sympy.sympify('x**2.0')

assert expr1.equals(expr2) 
assert expr1.equals(expr3) --> AssertionError (returns: None)

**Comment by asmeurer** [comment]
I think the evaluate=False part isn't relevant here. `(x**2).equals(x**2.0)` does the same. The float probably causes it to not evaluate to exactly 0, so it gives None because it is not sure. I'm not sure if this can, or even should, be fixed.

**Comment by Karrenbelt** [comment]
true: evaluate=False just keeps it x*x instead of evaluating to x**2

can you give me the reason why the float should not be equal?
```
assert sympy.sympify('x*2.0').equals(sympy.sympify('x*2'))
assert sympy.sympify('2.0*x').equals(sympy.sympify('2*x'))
```
works fine
`sympy.sympify('2.0**x').equals(sympy.sympify('2**x')) `
also doesn't

**Comment by asmeurer** [comment]
None doesn't mean they aren't equal (it would give False in that case). It means that it can't determine either way. `equals` works by first trying to simplify the expression, then substituting random values and seeing if they give non-zero. SymPy doesn't reduce `x**2 - x**2.0` to 0, so the first test gives inconclusive. On the other hand, any value that you substitute in that expression will give 0. Since substituting values can only conclusively prove that two expressions aren't equal, it gives None. 

I agree that `2*x - 2.0*x == 0` is inconsistent. There are lots of subtleties with this, and the current behavior isn't necessarily the ideal. A good starting point is the discussions here, as well as the cross-referenced issues https://github.com/sympy/sympy/issues/11707

**Comment by smichr** [comment]
> A good starting point is the discussions here, as well as

Continued kudos for the role of referene librarian that you play regularly! I appreciate these bread-crumbs.

**Comment by smichr** [comment]
@Karrenbelt , this might be a diff worth pursuing with a PR:
```diff
diff --git a/sympy/core/expr.py b/sympy/core/expr.py
index bd3cf4e..a6c7155 100644
--- a/sympy/core/expr.py
+++ b/sympy/core/expr.py
@@ -745,7 +745,9 @@ def equals(self, other, failing_expression=False):
         from sympy.polys.numberfields import minimal_polynomial

         other = sympify(other)
-        if self == other:
+        # first try the most trivial case
+        if self == other or (self.func == other.func and
+                self.args == other.args):
             return True

         # they aren't the same so see if we can make the difference 0;
```

**Comment by asmeurer** [comment]
> Continued kudos for the role of referene librarian that you play regularly! I appreciate these bread-crumbs.

It's helpful that I can remember these discussions, but I am finding it increasingly harder to find old issues with GitHub's search. The other day I spent probably 10 minutes trying to find a specific issue that I knew existed. Sometimes I just give up. I'm hoping either GitHub improves things or someone else builds something better. 

> @Karrenbelt , this might be a diff worth pursuing with a PR:

I'm not sure about this. `__eq__` will usually be defined exactly as func == func and args == args, and when it is defined differently, shouldn't equals respect that?

**Comment by smichr** [comment]
> I'm not sure about this

Ah...The difference simplifies even though the terms are not equal in the Mul case:
```python
>>> 2*x == 2.*x
False
>>> (2.*x-2*x)
0
>>> (2.**x-2**x)
-2**x + 2.0**x
```
So that's why the former give True from `equals` while the latter gives None.

**Comment by asmeurer** [comment]
There's probably some things that should be changed here a la the referenced issue, but I don't think the changes should be in equals. It's just uncovering the inconsistencies deeper in the core.

**Comment by smichr** [comment]
> inconsistencies deeper in the core.

With `S(2) - S(2.0) == 0` the way to fix this would be to restrict collection of terms in `Add.flatten` so that coefficients of different types don't cancel.

**Comment by asmeurer** [comment]
Isn't numeric constants always combining a fundamental behavior of Add and Mul? That seems like it would break a lot of things.

**Comment by smichr** [comment]
> would break a lot of things.

I doubt that we use mixed-type coefficients very often, but I could be wrong.
